src/R2N.jl

```diff
@@ -0,0 +1,412 @@
+
+    ΔTk = slope + curv / 2
+    ck .= x .+ s
+    fck = obj(nlp, ck)
+
+    if non_mono_size > 1  #non-monotone behaviour
+      k = mod(stats.iter, non_mono_size) + 1
+      solver.obj_vec[k] = stats.objective
+      fck_max = maximum(solver.obj_vec)
+      ρk = (fck_max - fck) / (fck_max - stats.objective + ΔTk)
+    else
+      ρk = (stats.objective - fck) / ΔTk
+    end
+
+    # Update regularization parameters and Acceptance of the new candidate
+    step_accepted = ρk >= η1 && σk >= η2
+    if step_accepted
+      μk = max(μmin, μk / λ)
+      x .= ck
+      grad!(nlp, x, ∇fk)
+      if isa(nlp, QuasiNewtonModel)
+        ∇fn .-= ∇fk
+        ∇fn .*= -1  # = ∇f(xₖ₊₁) - ∇f(xₖ)
+        push!(nlp, s, ∇fn)
+        ∇fn .= ∇fk
+      end
+      set_objective!(stats, fck)
+      unbounded = fck < fmin
+      norm_∇fk = norm(∇fk)
+    else
+      μk = μk * λ
+    end
+
+    set_iter!(stats, stats.iter + 1)
+    set_time!(stats, time() - start_time)
+
+    cgtol = max(rtol, min(T(0.1), √norm_∇fk, T(0.9) * cgtol))
+    set_dual_residual!(stats, norm_∇fk)
+
+    solver.σ = σk
+    solver.μ = μk
+    solver.cgtol = cgtol
+
+    callback(nlp, solver, stats)
+
+    norm_∇fk = stats.dual_feas # if the user change it, they just change the stats.norm 
+    μk = solver.μ
+    cgtol = solver.cgtol
+    σk = μk * norm_∇fk
+
+    optimal = norm_∇fk ≤ ϵ
+
+    if verbose > 0 && mod(stats.iter, verbose) == 0
+      @info log_row([
+        stats.iter,            # Current iteration number
+        stats.objective,       # Objective function value
+        norm_∇fk,              # Gradient norm
+        μk,                    # Mu value
+        σk,                    # Sigma value
+        ρk,                    # Rho value
+        step_accepted ? "↘" : "↗", # Step acceptance status
+      ])
+    end
+    if stats.status == :user
+      done = true
+    else
+      set_status!(
+        stats,
+        get_status(
+          nlp,
+          elapsed_time = stats.elapsed_time,
+          optimal = optimal,
+          unbounded = unbounded,
+          max_eval = max_eval,
+          iter = stats.iter,
+          max_iter = max_iter,
+          max_time = max_time,
+        ),
+      )
+      done = stats.status != :unknown
+    end
+  end
+
+  set_solution!(stats, x)
+  return stats
+end
+
+function subsolve!(R2N::R2NSolver, s, atol, n, subsolver_verbose)
+  ∇f = R2N.gx
+  cgtol = R2N.cgtol
+  H = R2N.H
+  subsolver_type = R2N.subsolver_type
+  σ = R2N.σ
+  opI = R2N.opI
+
+  if subsolver_type isa MinresSolver
+    minres!(
+      subsolver_type,
+      H,
+      ∇f, #b 
+      λ = σ,
+      itmax = 2 * n,
+      atol = atol,
+      rtol = cgtol,
+      verbose = subsolver_verbose,
+    )
+    s .= subsolver_type.x
+  elseif subsolver_type isa KrylovSolver
+    Krylov.solve!(
+      subsolver_type,
+      H + opI * σ,
+      ∇f,
+      atol = atol,
+      rtol = cgtol,
+      itmax = 2 * n,
+      verbose = subsolver_verbose,
+    )
+    s .= subsolver_type.x
```

Comment:
You have to check the status of the subsolver and make sure it solved successfully.